### ISSUE TITLE
Refactor tests to use Assert.Throws

### ISSUE THREAD

**Issue by nightblade9** [Chris3606/GoRogue#145]
While browsing the code for effects, I noticed [this little gem](https://github.com/Chris3606/GoRogue/blob/84664c66b6bfcfd2eed345d7fc6d21da2204df30/GoRogue-UnitTests/EffectTests.cs#L74). It's a try/catch block that sets a value to `true`, then uses `Assert.AreEqual(blah, true)`.

You can replace this entire block:

```
			bool exception = false;
			try
			{
				effectTrigger.Add(new IntEffect("Test Effect 0", 0));
			}
			catch (ArgumentException)
			{
				exception = true;
			}
                        Assert.AreEqual(true, exception);
```

With something simpler/cleaner:

```
Assert.Throws<ArgumentException>(() => effectTrigger.Add(new IntEffect("Test Effect 0", 0)));
```

I personally prefer the latter since it's more readable and clear. It also fails if, say, the wrong exception type throws instead of the expected type.

You can also assign the value to a variable and assert more things, if you like; something like:

```
var ex = Assert.Throws<...>(...)
Assert.That(ex.CustomMessageProperty, Is.EqualTo(foo))
```

**Comment by Chris3606** [comment]
LOL.... I'm so sorry, I have no idea what would have possessed me to code that the way I did.  Although both should be equivalent (in that the first should also fail if the wrong exception is thrown), however the second is clearly more readable and better expresses the intent.

I won't likely be able to fix this for a bit, as I likely won't be near a reasonable dev environment for about a week, but as soon as I'm able I'll fix this.

Fair warning, granted most are not this bad, but GoRogue's unit tests are in general woefully inadequate.  The cases they test for are at best incomplete, and as well many of the unit cases don't have proper asserts and are just basically "print statements" whose output is to be viewed manually (unit tests like this have names that start with "Manual").

There's been a backlog item on this issues page for quite awhile to clean them up, but its proven to be enough work that it hasn't managed to get done.  Suffice it to say, I wouldn't consider them a terribly accurate or even complete representation of the "proper" way to use GoRogue features at current time.

**Comment by nightblade9** [comment]
That's interesting .... I am not able to find information about some features (such as path-finding or triggers/events) and resort to reading the tests, thinking (as someone exposed to TDD) that these represent the use-cases for how methods work.

Since you can't solve your test problem overnight, the best place to start is to write better tests for new things. 

I'll see if I can poke around the docs and/or write blog posts about things as I figure them out - maybe that will help.

**Comment by Chris3606** [comment]
TLDR; See docs header below for some pointers to helpful resources, I'm also more than happy to answer questions here in github issues or 1-on-1.

What you're thinking of (TDD-oriented architecture) was the original intent, but that wasn't well maintained and proved to work extremely poorly for parts of the project like FOV where the feedback was pretty much all visual as to whether it's "working" correctly.  It's the same problem with using TDD for UI-based apps, it's more work to fake the UI than to do something else.  In fact, its actually worse than the UI problem for GoRogue because there isn't one correct answer for many algorithms and there is RNG involved.  As such, the unit tests are helpful now but aren't meant to be feature-complete.

A lot of the tests were accurate to the preferred way of doing things when they were written, and while the method the tests use is still functional today, the preferred way of doing things has changed significantly thanks to feature additions.  For example, many of the tests custom-implement a map view, when with today's feature set a 1-line LambdaMapView or some such is all that is necessary.  Particularly since I found the tests to be of limited use for some of the features, they simple haven't been updated.

# Documentation
Documentation wise, while the API docs are complete, there are some features that simply aren't documented in detail at the moment other than the API docs.  I recommend checking [this link](https://chris3606.github.io/GoRogue/) -- it contains the API documentation and some articles that further explain some features, provide examples, etc.  The API documentation is complete and detailed, and while the articles section is nowhere near as complete feature-wise as I would like, what is there should be accurate to the current version and more or less complete for the feature documented.

There is also the [GitHub wiki](https://github.com/Chris3606/GoRogue/wiki), which has some additional articles, however the GitHub wiki is being phased out currently in favor of the first link, and what's on the wiki isn't in a lot of cases updated.  I'm currently in the process of updating/porting what is on the wiki to the first link.  What is on the wiki won't probably be updated further at that location, and eventually the wiki will be removed (or will simply point to the first link).

As well, if there's any questions or confusion you have, I'm more than happy to answer them directly, feel free to ask.  You can either post a GitHub issue, post on the GoRogue subreddit [r/goroguelib](www.reddit.com/r/goroguelib), or feel free to contact me on Discord -- whatever you prefer.  I'm also actively in the process of resuming work on updating the docs at the link I shared above, so if there's a particular feature you're interested in, I can certainly try to prioritize it.

Good luck!
